File: mobile/ios/PrestamosApp/Services/SyncService.swift
import Foundation
import Alamofire

/**
 * Servicio de sincronización offline para iOS
 */
class SyncService {
    private let offlineHelper = OfflineHelper()
    private let apiService = ApiService.shared
    
    /**
     * Sincronizar todas las acciones pendientes
     */
    func sincronizar() {
        let acciones = offlineHelper.obtenerAccionesPendientes()
        
        if acciones.isEmpty {
            print("SyncService: No hay acciones pendientes")
            return
        }
        
        print("SyncService: Sincronizando \(acciones.count) acciones pendientes")
        
        for accion in acciones {
            do {
                try sincronizarAccion(accion)
                offlineHelper.eliminarAccionPendiente(accion)
            } catch {
                print("SyncService: Error sincronizando acción: \(error.localizedDescription)")
                offlineHelper.incrementarIntentos(accion)
                
                // Si tiene más de 5 intentos, eliminarla
                if accion.intentos >= 5 {
                    offlineHelper.eliminarAccionPendiente(accion)
                }
            }
        }
        
        offlineHelper.guardarUltimaSincronizacion()
    }
    
    /**
     * Sincronizar una acción específica
     */
    private func sincronizarAccion(_ accion: OfflineHelper.PendingAction) throws {
        switch accion.tipo {
        case "crear_prestamo":
            guard let data = accion.datos.data(using: .utf8),
                  let request = try? JSONDecoder().decode(PrestamoRequest.self, from: data) else {
                throw NSError(domain: "SyncService", code: -1, userInfo: [NSLocalizedDescriptionKey: "Error decodificando datos"])
            }
            
            apiService.createPrestamo(request) { result in
                switch result {
                case .success:
                    print("SyncService: Préstamo sincronizado")
                case .failure(let error):
                    throw error
                }
            }
            
        case "crear_cliente":
            guard let data = accion.datos.data(using: .utf8),
                  let request = try? JSONDecoder().decode(ClienteRequest.self, from: data) else {
                throw NSError(domain: "SyncService", code: -1, userInfo: [NSLocalizedDescriptionKey: "Error decodificando datos"])
            }
            
            apiService.createCliente(request) { result in
                switch result {
                case .success:
                    print("SyncService: Cliente sincronizado")
                case .failure(let error):
                    throw error
                }
            }
            
        case "registrar_pago":
            guard let data = accion.datos.data(using: .utf8),
                  let request = try? JSONDecoder().decode(PagoRequest.self, from: data) else {
                throw NSError(domain: "SyncService", code: -1, userInfo: [NSLocalizedDescriptionKey: "Error decodificando datos"])
            }
            
            apiService.registrarPago(request) { result in
                switch result {
                case .success:
                    print("SyncService: Pago sincronizado")
                case .failure(let error):
                    throw error
                }
            }
            
        default:
            print("SyncService: Tipo de acción desconocido: \(accion.tipo)")
        }
    }
    
    /**
     * Verificar conectividad y sincronizar si hay conexión
     */
    func sincronizarSiHayConexion() {
        let manager = NetworkReachabilityManager()
        
        if manager?.isReachable ?? false {
            DispatchQueue.global(qos: .background).async {
                self.sincronizar()
            }
        } else {
            print("SyncService: Sin conexión, no se puede sincronizar")
        }
    }
}

File: mobile/android/app/src/main/java/com/erp/prestamos/services/SyncService.java
package com.erp.prestamos.services;

import android.content.Context;
import android.util.Log;
import com.erp.prestamos.utils.OfflineHelper;
import com.erp.prestamos.models.*;
import com.google.gson.Gson;
import java.util.List;

/**
 * Servicio de sincronización offline
 */
public class SyncService {
    private static final String TAG = "SyncService";
    private Context context;
    private ApiServiceHelper apiHelper;
    private OfflineHelper offlineHelper;
    private Gson gson;
    
    public SyncService(Context context) {
        this.context = context;
        this.apiHelper = ApiServiceHelper.getInstance(context);
        this.offlineHelper = new OfflineHelper(context);
        this.gson = new Gson();
    }
    
    /**
     * Sincronizar todas las acciones pendientes
     */
    public void sincronizar() {
        List<OfflineHelper.PendingAction> acciones = offlineHelper.obtenerAccionesPendientes();
        
        if (acciones.isEmpty()) {
            Log.d(TAG, "No hay acciones pendientes para sincronizar");
            return;
        }
        
        Log.d(TAG, "Sincronizando " + acciones.size() + " acciones pendientes");
        
        for (OfflineHelper.PendingAction accion : acciones) {
            try {
                sincronizarAccion(accion);
                offlineHelper.eliminarAccionPendiente(accion);
            } catch (Exception e) {
                Log.e(TAG, "Error sincronizando acción: " + e.getMessage());
                offlineHelper.incrementarIntentos(accion);
                
                // Si tiene más de 5 intentos, eliminarla
                if (accion.intentos >= 5) {
                    offlineHelper.eliminarAccionPendiente(accion);
                }
            }
        }
        
        offlineHelper.guardarUltimaSincronizacion();
    }
    
    /**
     * Sincronizar una acción específica
     */
    private void sincronizarAccion(OfflineHelper.PendingAction accion) throws Exception {
        switch (accion.tipo) {
            case "crear_prestamo":
                PrestamoRequest prestamoRequest = gson.fromJson(accion.datos, PrestamoRequest.class);
                apiHelper.createPrestamo(prestamoRequest, new ApiServiceHelper.CreatePrestamoCallback() {
                    @Override
                    public void onSuccess(Prestamo prestamo) {
                        Log.d(TAG, "Préstamo sincronizado: " + prestamo.getId());
                    }
                    
                    @Override
                    public void onError(String error) {
                        throw new RuntimeException(error);
                    }
                });
                break;
                
            case "crear_cliente":
                ClienteRequest clienteRequest = gson.fromJson(accion.datos, ClienteRequest.class);
                apiHelper.createCliente(clienteRequest, new ApiServiceHelper.CreateClienteCallback() {
                    @Override
                    public void onSuccess(Cliente cliente) {
                        Log.d(TAG, "Cliente sincronizado: " + cliente.getId());
                    }
                    
                    @Override
                    public void onError(String error) {
                        throw new RuntimeException(error);
                    }
                });
                break;
                
            case "registrar_pago":
                PagoRequest pagoRequest = gson.fromJson(accion.datos, PagoRequest.class);
                apiHelper.createPago(pagoRequest, new ApiServiceHelper.CreatePagoCallback() {
                    @Override
                    public void onSuccess(Pago pago) {
                        Log.d(TAG, "Pago sincronizado: " + pago.getId());
                    }
                    
                    @Override
                    public void onError(String error) {
                        throw new RuntimeException(error);
                    }
                });
                break;
                
            case "crear_ruta":
                RutaRequest rutaRequest = gson.fromJson(accion.datos, RutaRequest.class);
                apiHelper.createRuta(rutaRequest, new ApiServiceHelper.CreateRutaCallback() {
                    @Override
                    public void onSuccess(Ruta ruta) {
                        Log.d(TAG, "Ruta sincronizada: " + ruta.getId());
                    }
                    
                    @Override
                    public void onError(String error) {
                        throw new RuntimeException(error);
                    }
                });
                break;
                
            default:
                Log.w(TAG, "Tipo de acción desconocido: " + accion.tipo);
        }
    }
    
    /**
     * Verificar conectividad y sincronizar si hay conexión
     */
    public void sincronizarSiHayConexion() {
        android.net.ConnectivityManager cm = 
            (android.net.ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        
        android.net.NetworkInfo activeNetwork = cm.getActiveNetworkInfo();
        boolean isConnected = activeNetwork != null && activeNetwork.isConnectedOrConnecting();
        
        if (isConnected) {
            new Thread(() -> sincronizar()).start();
        } else {
            Log.d(TAG, "Sin conexión, no se puede sincronizar");
        }
    }
}

